api/x509cert.go

```diff
@@ -108,10 +112,33 @@ func (s *SigningService) PostX509Certificate(ctx context.Context, request *proto
 		return nil, status.Errorf(codes.InvalidArgument, "Bad request: %v", err)
 	}
 
-	data, err := s.SignX509Cert(req, request.KeyMeta.Identifier)
-	if err != nil {
+	type resp struct {
+		data []byte
+		err  error
+	}
+	respCh := make(chan resp)
+	go func() {
+		data, err := s.SignX509Cert(reqCtx, req, request.KeyMeta.Identifier)
+		respCh <- resp{data, err}
+	}()
+
+	select {
+	case <-ctx.Done():
+		// client canceled the request. Cancel any pending server request and return
+		cancel()
 		statusCode = http.StatusInternalServerError
-		return nil, status.Error(codes.Internal, "Internal server error")
+		err = fmt.Errorf("client canceled request for %q", config.X509CertEndpoint)
+		return nil, status.Errorf(codes.Canceled, "%v", err)
+	case <-reqCtx.Done():
+		// server request timed out.
+		statusCode = http.StatusGatewayTimeout
```

Comment:
`http.StatusServiceUnavailable` would be more appropriate - https://tools.ietf.org/html/rfc7231#section-6.6.4